sogs/config.py

```diff
@@ -1,5 +1,126 @@
-from .base_config import *
+import configparser
+import os
+import re
+import logging
+import coloredlogs
 
-# Add any config overrides here.
-# (If running out of a git checkout you may want `git update-index --skip-worktree sogs/config.py`
-# to make git ignore changes you make here).
+from .web import app
+
+# Set up colored logging; we come back to set the level once we know it
+coloredlogs.install(level='NOTSET', milliseconds=True, isatty=True)
+
+# Default config settings; most of these are configurable via config.ini (see it for details).
+DB_PATH = 'sogs.db'
+DB_SCHEMA_FILE = 'schema.sql'
+KEY_FILE = 'key_x25519'
+URL_BASE = 'http://example.net'
+HTTP_SHOW_RECENT = True
+OMQ_LISTEN = 'tcp://*:22028'
+OMQ_INTERNAL = 'ipc://./omq.sock'
+LOG_LEVEL = 'WARNING'
+UPLOAD_DEFAULT_EXPIRY_DAYS = 15
+UPLOAD_FILENAME_MAX = 60
+UPLOAD_FILENAME_KEEP_PREFIX = 40
+UPLOAD_FILENAME_KEEP_SUFFIX = 17
+UPLOAD_FILE_MAX_SIZE = 6_000_000
+UPLOAD_FILENAME_BAD = re.compile(r"[^\w+\-.'()@\[\]]+")
+ROOM_ACTIVE_PRUNE_THRESHOLD = 60
+ROOM_DEFAULT_ACTIVE_THRESHOLD = 7
+MESSAGE_HISTORY_PRUNE_THRESHOLD = 30
+IMPORT_ADJUST_MS = 0
+BAD_WORDS_FILE = 'badwords.txt' if os.path.exists('badwords.txt') else None
+
+
+def load_config():
+    if 'SOGS_CONFIG' in os.environ:
+        conf_ini = os.environ['SOGS_CONFIG']
+        if conf_ini and not os.path.exists(conf_ini):
+            raise RuntimeError(f"SOGS_CONFIG={conf_ini} specified, but path does not exist!")
+    else:
+        conf_ini = 'sogs.ini'
+        if not os.path.exists(conf_ini):
+            app.logger.info("sogs.ini does not exist; using all config defaults")
+            conf_ini = None
+
+    if not conf_ini:
+        return
+
+    app.logger.info(f"Loading config from {conf_ini}")
+    cp = configparser.ConfigParser()
+    cp.read(conf_ini)
+
+    # Set log level up first (we'll set it again below, mainly to log it if we have debug logging
+    # enabled).
+    if 'log' in cp.sections() and 'level' in cp['log']:
+        app.logger.setLevel(cp['log']['level'])
+
+    # Map of: section => { param => ('GLOBAL', test lambda, value lambda) }
+    # global is the string name of the global variable to set
+    # test lambda returns True/False for validation (if None/omitted, accept anything)
+    # value lambda extracts the value (if None/omitted use str value as-is)
+    setting_map = {
+        'db': {'url': ('DB_PATH', lambda x: x.startswith('sqlite:///'), lambda x: x[10:])},
```

Comment:
should it be `sqlite://` instead here?

Reply:
No, the triple slashes is correct (for future SQL alchemy url support).  Formally it is `dialect://host/stuff` but sqlite has no host.

Reply:
(and for an absolute path it's `sqlite:////full/path`)

Reply:
ew, ok.